### DISCUSSION TITLE
A feature that I didn't find to exist, please educate if it does; please otherwise consider its need

### DISCUSSION THREAD

**Discussion by rajibando** [zim-desktop-wiki/zim-desktop-wiki#2599] [General]
Hello, the Team Leaders and the Senior Members of Zim-Wiki github discussions group,

Please accept my admiration for having such a wonderful software up and running, which I have been using for the last 3 months. I like it for a particular personal private project of my own.

Yesterday, I had a chat at https://web.libera.chat/#zim-wiki with Mr. Barlas. I thank him for his support. My hunch was reinforced that while using Zim, within a zim document spanning multiple pages, I can't link a particular phrase in one of the pages of the said multi-page zim document to another phrase in another page of the said same multi-page zim document, unless I use the Heading  tool.

Perhaps a tool that uses an indexing tool could help in this regard, only that we could enter phrases as well as words for their indexing.

My personal, private, note-taking project is rudimentary and I didn't require much skill in creating, and proceeding with, the project. So please note that my skill with zim desktop wiki is very rudimentary. I would welcome any lesson that might appear apt for my need.

This particular tool should be a boon. We could travel from one particular portion of one page to another portion of another page containing similar ideas, using these links, without using the Heading object.

I could later use an external indexing tool within the folder to capture strings used within the multi-page document and use the selected ones for the purpose of indexing later on.

In a multi-page zim desktop wiki document, there could be multiple instances of one particular idea using one key phrase. With my imagined tool, the idea could turn out to be important.

I hope that the zim team will educate me, if such a tool exists, or look into this suggestion, if it doesn't.

Best wishes,
Rajib Bandopadhyay
Etc.

**Comment by Nick222** [comment]
It seems to me that what is needed here is a subject index tool that would allow one to move from one occurrence of a term to another (not directly, but through a special list of word terms).
At the same time, there really is a problem of connection between pieces of text on the same page, but I think that it can be solved either by dividing one note into many small ones, or through tags that could be assigned to each individual paragraph.

**Reply by rajibando** [reply]
Thank you, Mr. Nick222, for your post, suggestions and initiating a discussion.
Could you please reconsider the phrases you used, "subject index tool" and "a special list of word terms"? It should theoretically be "phrase index tool" and for "any phrase" respectively.
Regarding the portion of your comment, "... there really is a problem of connection between pieces of text on the same page ...", it would be connections between pieces of text on different pages within the same document.
So you might need to amend the subsequent portion of your statement,  "...  think that it can be solved either by dividing one note into many small ones, or through tags that could be assigned to each individual paragraph ...".
I can say this much for your "... by dividing one note into many small ones ..." such break up would cause the tight integrity to be lax.
Tag is something that I have to learn using.
The Preferences/Plugins/Tags say, "... plugin provides a page index filtered by means of selecting tags in a cloud ...". Regarding "tags in a cloud"? Didn't make sense.
I would like to learn more about this wonderful tool to use it better. But visually, the tag with "@" appears as an eyesore in plain text pages.
It would be nice is the "@" be replaced by some much lesser conspicuous, on invisible, icons that remain visible as Unifont but invisible in ASCII rendering.

**Comment by introt** [comment]
Hi!

Have you come across **anchors**, described under Links in the manual?

> Finally there is a special "anchor" object to place a link anywhere in the text or to e.g. set an explicit id for a heading. This can be inserted by typing "''##name''" and is rendered like this: {{id: name}} .

https://zim-wiki.org/manual/Help/Links.html

**Reply by sojusnik** [reply]
The manual is very vague about the usage of these special "anchor" objects. Can you give some examples how to place such anchors within a text and link to it from the same ando another note?

**Comment by rajibando** [comment]
> Hi!
> 
> Have you come across **anchors**, described under Links in the manual?
> 
> > Finally there is a special "anchor" object to place a link anywhere in the text or to e.g. set an explicit id for a heading. This can be inserted by typing "''##name''" and is rendered like this: {{id: name}} .
> 
> https://zim-wiki.org/manual/Help/Links.html

Thank you, Mr. Introt, for your input. Unfortunately, I use an older Debian installed from `cdrom:[Official Debian GNU/Linux Live 11.6.0 lxde 2022-12-17T11:46]` and an older Zim. 
$zim --version
zim 0.73.5.

So anchor isn't in the help file. Did the version `zim 0.73.5.` contain the tool Anchors?
I believe, No.

Later addition:
I downloaded zim 0.75.2 from Debian `trixie` and checked that all dependencies were satisfied for my system. So I downloaded and installed it. It is currently running. So I shall proceed with the anchor tool and get back.

Bottomline:
Ok, the text is understood: 

> This can be inserted by typing "''##name''" and is rendered like this: {{id: name}} .

But how is the id: name" accessed?

Suppose there are two text docs within the top level zim document notebook.zim. Say one is read.txt and the other is write.txt

**read.txt**

> 'Without' reading skills writing skills can't develop. So  read and write . Or remain functionally illiterate. Reading the script is a must for transferring thoughts into writing. Writing is a great tool. It needs to be nurtured. So read and write, otherwise you can't develop the skill of writing.


**write.txt**

> After one has learnt to read, one must learn the habit to write, to form thoughts and try writing small passages. Then try reading it. So read and write.

In the read.txt file, we have two instances of the word "write" and two instances of the word "read". Please use the first instance for normal link to write.txt file. Use the 2nd instance to link to the second "write" in the write.txt via anchor.

Similarly, in the write.txt file, we have two instances of "read" and two instances of the word "write". Please use the first instance for normal link to read.txt file. Use the 2nd instance to link to the second "write" in the write.txt via anchor.. Use the first instance for normal link to read.txt. Use the 2nd instance to link to the second "read" in read.txt via anchor.

Now suppose that I want to access the the **anchor** Ids  read from write.txt and write from read.txt?

This passage will form an apt example for illustrating the usage of **Anchor** object with "##".

**Reply by introt** [reply]
Yes, anchors were introduced in 0.74. After typing `##name`and a space or a newline, the anchor is created. It seems they're not rendered properly in the manual, they should look like this: ![image](https://github.com/zim-desktop-wiki/zim-desktop-wiki/assets/26069787/b8158acf-b037-483b-aea0-850cb307ca40)

You can click on the anchor to reveal its id and to copy a link to it:

![image](https://github.com/zim-desktop-wiki/zim-desktop-wiki/assets/26069787/5ece6c67-38a8-4dfa-b134-d4fe340154b1)

The link can then be pasted elsewhere. The link will open the page at the position of the anchor.

The `##name` form is how you create an anchor, while the actual wiki syntax, visible in the page source, is `{{id: name}}`. Thus your example would look something like this:

![image](https://github.com/zim-desktop-wiki/zim-desktop-wiki/assets/26069787/38c88555-9aca-4168-93bc-37aaedc2785b)

```
'Without` reading skills writing skills can't develop. "[[write#write"]]. Or remain functionally illiterate. Reading the script is a must for transferring thoughts into writing. Writing is a great tool. It needs to be nurtured. Write, otherwise you can't develop the skill of writing. {{id: read}}
```

![image](https://github.com/zim-desktop-wiki/zim-desktop-wiki/assets/26069787/ff66cabe-8355-415a-90ca-ed602683402f)

```
After one has learnt to [[read#read]], one must learn the habit to form thoughts and try writing small passages. Then [[read#read|read]] it. {{id: write}}
```

I set the link text of the second link in `write.txt` to "read" - the default pasted form is `page name#id`.

**Reply by rajibando** [reply]
Mr. Introt, a huge thank you. I am getting there. So please stay with me.
I had committed a mistake in the write.txt by putting two ##read. I have removed the second ##read, which I had overlooked.
I then corrected the two files and made the two anchors and links. What I get is this:
I have to begin with a blank space, write a word, put two hashes, write the word again, then insert a space. The anchor and icon is created. I then go back and remove the redundant spaces that were inserted to create the anchor and its icon. A little awkward.
May be, in the next edition, a tool on the toolbar and a combo-keystrokes could be created to remove these extra spaces after the anchor creation. 
You may choose to edit your post likewise.
I have tested. I am getting there. So please hang on to this topic. The files are here: 
[ZimAnchor.tar.gz](https://github.com/zim-desktop-wiki/zim-desktop-wiki/files/14970580/ZimAnchor.tar.gz)
Also, please request the moderators to use my above lines in the Manual file to explain **Anchor**.  Like this:
_To create an anchor with icon and Id., begin with a blank space, write a word, insert a space, put two hashes, write the word again, then insert a space. The anchor and icon is created. Then go back and remove the redundant spaces that were inserted to create the anchor and its icon._

I will subsequently try to check if phrase anchors could be created.

**Reply by sojusnik** [reply]
@rajibando Did you enable "convert hashtags to links" in Zim's settings? That's what was missing in my case.

**Reply by introt** [reply]
Sorry, I'm not following - you seem to have removed the anchors from your example completely. Neither do I understand why you should "write the word again". This is what the process should look like:

[anchor.webm](https://github.com/zim-desktop-wiki/zim-desktop-wiki/assets/26069787/a3d0df56-24b7-4622-b1aa-bcc8c6316e2f)

I personally don't mind the space before the anchor, and instead of a space afterwards, you may use a line break. You might want to use a custom tool for getting rid of extra spaces around anchors - a keybinding can be added to run those, and I do believe they can also be shown in the toolbar if the toolbar plugin is enabled.

Regarding "phrase anchors", underscores and dashes should do the trick. This could also be made more clear in the docs. I've labeled this discussion appropriately, am not really in a doc writing mood rn.

**Reply by rajibando** [reply]
Mr. Introt, did you download the compressed files, put them up in a directory and checked? I have not removed the anchors. I have modified and optimised them.
You needn't have uploaded the webm video. I had understood and implemented the idea already. Just took some time to do it appropriately.
Speaking of which, for ##string to be converted into an anchor automatically, do we need to activate an add-on/tool?
May please the concerned files be checked:
[write.txt](https://github.com/zim-desktop-wiki/zim-desktop-wiki/files/14971461/write.txt)
[read.txt](https://github.com/zim-desktop-wiki/zim-desktop-wiki/files/14971462/read.txt)
[Home.txt](https://github.com/zim-desktop-wiki/zim-desktop-wiki/files/14971463/Home.txt)
Please advise with your inputs.
Then with specific examples could you elaborate on the following?:
> ... Regarding "phrase anchors", underscores and dashes should do the trick ...

How do you highlight a phrase as a link without displaying the "_" or "-"? I can understand the idea like "long phrase" as ##long_phrase, or "long phrase" as ##long-phrase, without the "". But getting "long phrase" (without "") highlighted in full as a link with an anchor? How?

**Reply by introt** [reply]
Ah sorry, I referred to your post.

> for ##string to be converted into an anchor automatically, do we need to activate an add-on/tool?

No, but you do need to have "convert hashtags to links" enabled as Sojusnik mentioned.

> Please advise with your inputs.

I feel the anchors are a bit redundant in this example, as the page title headers have the same ids as the anchors - their natural habitat lies in longer text so they're a bit hard to demo with small snippets.

> How do you highlight a phrase as a link without displaying the "_" or "-"?

I'm assuming you're alluding to the link text? `[[target_page#target_anchor_or_heading|Link text]]` The anchor's id isn't really supposed to be seen by the reader.

**Reply by rajibando** [reply]
Ok, thanks, Mr. [Introt](https://github.com/introt). Apologies for the belated reply. I realised that I should have thanked you, when I perused the discussion today again.

**Reply by rajibando** [reply]
And Mr. [Introt](https://github.com/introt) on hindsight, my zipped zim desktop wiki example  files and the directory, as my first tentative steps towards learning about `Anchor` was in  fact a very poor representation.
However, we learn by making mistakes. So apologies for the confusion I created, if any, with my said stupid examples on `Anchor`.
Thank you for introducing me to `Anchor`.

**Comment by frogmouse** [comment]
So glad to find this discussion.  The [current documentation](https://zim-wiki.org/manual/Help/Links.html) says this:  

> 
Finally there is a special "anchor" object to place a link anywhere in the text or to e.g. set an explicit id for a heading. This can be inserted by typing "##name" and is rendered like this: .  

Note there's nothing between the colon and the period.

So now that I think I know how to use anchors., I'll have to go try them out.
